#12516: Implementing a Barrier CCL OP

jvegaTT opened the pull request:
### Ticket
Link to Github Issue
https://github.com/tenstorrent/tt-metal/issues/12516

### Problem description
Provide context for the problem.

New feature - a barrier operation to synchronize all the devices in a ring configuration, assures all previous OPs on all devices has completed before proceeding and initializes them at the same time

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

Send a message around all elements in a ring that verifies that all  previous scheduled operations on that device has completed. Upon completion of the loop sends a synchronization message which then allows the next scheduled operation to run on the device

Added t3k frequent test to verify the above functionality

Applied bug fix to ethernet latency microbenchmark unit test where the semaphore address was incorrect and the test was writing the semaphore to the allocated buffer thereby corrupting the data, and was delaying the start.

Applied enhancements to Ethernet API

Minor corrections to tech report

### Checklist
- [X] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/11422565483 -- reran after suggested changes applied here: https://github.com/tenstorrent/tt-metal/actions/runs/11560035556
- [ ] Blackhole Post commit (if applicable) Not applicable, only implemented for wormhole t3k
- [x] Model regression CI testing passes (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/11410998793 -- reran after suggested changes applied here: https://github.com/tenstorrent/tt-metal/actions/runs/11595168359
- [ ] Device performance regression CI testing passes (if applicable) OP not yet recommended for performance to ML team, for use with internal testing
- [X] New/Existing tests provide coverage for changes Added t3k frequent test suite for barrier

jvegaTT committed:
#12516 Created non-device connection for barrier implementation

jvegaTT committed:
#12516 Created non-device connection for barrier implementation

jvegaTT committed:
#12516: Made ping latency uBenchmark into OP

jvegaTT committed:
#12516: Made ping latency uBenchmark into OP

jvegaTT committed:
#0: Bug fix on hop latencies uBenchmark

Runtime arguments were flipped between semaphore and Buffer

jvegaTT committed:
#0: Bug fix on hop latencies uBenchmark

Runtime arguments were flipped between semaphore and Buffer

jvegaTT committed:
#0: Update Programming Mesh of Devices with TT-NN.md

jvegaTT committed:
#0: Update Programming Mesh of Devices with TT-NN.md

jvegaTT committed:
#12516: Added barrier op, has ethernet semaphore issue

Adding the barrier operation but it fails due to the fact ethernet semaphore does not initialize correctly

jvegaTT committed:
#12516: Added barrier op, has ethernet semaphore issue

Adding the barrier operation but it fails due to the fact ethernet semaphore does not initialize correctly

jvegaTT committed:
#0: Merge branch 'jvega/barrier' of github.com:tenstorrent/tt-metal into jvega/barrier

jvegaTT committed:
#12516: First working barrier
Need to gauge performance and effectiveness

jvegaTT committed:
#12516: First working barrier
Need to gauge performance and effectiveness

jvegaTT committed:
#0: adding the line that produces the trace_region_size bug

jvegaTT committed:
#12516: Added nightly and post commit testing to the barrier op

jvegaTT committed:
#12516: Moving barrier testing from nightly to t3000 frequent

jvegaTT committed:
#0: Correction to Bug fix on hop latencies uBenchmark commit

Also rotating the semaphore in the sender, not just in the receiver

jvegaTT committed:
#0: Adding license information to barrier_sem_creator.cpp

jvegaTT committed:
#0: license information added to barrier_full_worker_grid.cpp

jvegaTT committed:
#0: Fixing licensing bug

jvegaTT committed:
Merge branch 'main' into jvega/barrier

jvegaTT committed:
Merge branch 'main' into jvega/barrier

SeanNijjar commented on tests/ttnn/unit_tests/operations/test_barrier.py:
We really don't need all these tests for barrier (or the duplicated code from test all-gather).

I think we really only need 4 test (cases):
1) A simple barrier test running in a loop of some number of iteration (e.g. 1000) for 8 chips, 1 link
2) Above but 4 chips, 2 links (this will require a separate test function with the `pcie_mesh_device` fixture instead
3) (non-trace mode) A barrier + all-gather sequence test, run in a loop (e.g. 1000 iterations)
4) (trace mode) same as above but in trace mode

For the all-gather tests, we should just import functions like `is_unsupported_case` from the `test_all_gather` file and call directly. Realistically, we should in this case actually just assert that `is_unsupported_case` returns false because these are very hand-picked cases.

In terms of chosen all-gather case, it doesn't _really_ matter but choosing {chips: 8, links: 1, input_shape: [1,1,32,8192], dim: 3, dtype: bfp8} will reflect one of the llama configs we care about.

jvegaTT replied:
resolved in test_barrier_t3000_frequent.py

SeanNijjar commented on ttnn/cpp/pybind11/operations/__init__.hpp:
```diff
@@ -83,10 +89,6 @@ void py_module(py::module& module) {
     auto m_complex_unary_backward = module.def_submodule("complex_unary_backward", "complex_unary_backward operations");
     complex_unary_backward::py_module(m_complex_unary_backward);
 
-    auto m_ccl = module.def_submodule("ccl", "collective communication operations");
```
Please help preserve history readability hear by moving the original lines back to their original place. We should only see a 1-line diff (addition) for this file

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/barrier_sender_HPerf_LPrec.cpp:
```diff
@@ -0,0 +1,100 @@
+static constexpr bool DISABLE_CONTEXT_SWITCHING = true;
+static constexpr uint8_t NUM_CHANNELS = 8;
+
+FORCE_INLINE void eth_setup_handshake(uint32_t handshake_register_address) {
```
Same handshake comment as from receiver (but you can use the sender side functions instead)

TT-BrianLiu commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/barrier_receiver_HPerf_LPrec.cpp:
```diff
@@ -0,0 +1,180 @@
+
+#include <cstdint>
+#include "debug/assert.h"
+#include "debug/dprint.h"
```
Remove

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/barrier_sender_HPerf_LPrec.cpp:
```diff
@@ -0,0 +1,100 @@
+}
+
+FORCE_INLINE
+void eth_send_bytes_over_channel_payload_only(
```
Same comment as for receiver side eth dataflow like function

TT-BrianLiu commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/barrier_sender_HPerf_LPrec.cpp:
```diff
@@ -0,0 +1,100 @@
+#include <cstdint>
+#include "tt_metal/hw/inc/dataflow_api.h"
+#include "debug/assert.h"
+#include "debug/dprint.h"
```
Remove if you don't need

TT-BrianLiu commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/ubench_coordination.cpp:
```diff
@@ -0,0 +1,28 @@
+
+#include <cstdint>
+#include "debug/assert.h"
+#include "debug/dprint.h"
```
Remove if you don't need

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/barrier_sender_HPerf_LPrec.cpp:
```diff
@@ -0,0 +1,100 @@
+    uint32_t channels_sem_addrs, uint32_t ring_index)
+{
+    volatile uint32_t *sem_addr = reinterpret_cast<volatile uint32_t*>(channels_sem_addrs);
+    eth_noc_semaphore_wait(sem_addr, 1);
```
This function will eagerly context switch. You'll need to add idle timers here. I think we should add an idle time argument to the function just like I did with the handshake code in the handshake header.

TT-BrianLiu reviewed (changes requested):
Please switch new files to 2024.

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/barrier_sender_HPerf_LPrec.cpp:
```diff
@@ -0,0 +1,100 @@
+    uint32_t ring_index)
+{
+    uint32_t buffer_addr = channels_addrs;
+    eth_send_bytes_over_channel_payload_only(buffer_addr, buffer_addr, page_size + sizeof(eth_channel_sync_t), eth_channel_syncs);
```
loop while `!eth_txq_is_busy()` before calling (this goes with any eth call). Wait for some time before context switching.

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/barrier_sender_HPerf_LPrec.cpp:
```diff
@@ -0,0 +1,100 @@
+{
+    uint32_t buffer_addr = channels_addrs;
+    eth_send_bytes_over_channel_payload_only(buffer_addr, buffer_addr, page_size + sizeof(eth_channel_sync_t), eth_channel_syncs);
+    while (!eth_is_receiver_channel_send_done(0)) {
```
Same

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/barrier_sender_HPerf_LPrec.cpp:
```diff
@@ -0,0 +1,100 @@
+    channels_syncs_addrs->receiver_ack = 0;
+    *(volatile uint32_t*)channels_sem_addrs = 0;
+
+    //for (uint32_t i = 0; i < max_concurrent_samples; i++) {
```
delete

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/barrier_sender_HPerf_LPrec.cpp:
```diff
@@ -0,0 +1,100 @@
+    const bool is_ring_start = get_arg_val<uint32_t>(arg_idx++) == 1;
+    const uint32_t handshake_addr = get_arg_val<uint32_t>(arg_idx++);
+    const uint32_t ring_index = get_arg_val<uint32_t>(arg_idx++);
+    uint32_t channels_addrs = get_arg_val<uint32_t>(arg_idx++);
```
const where applicable

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/ubench_sender.cpp:
Is this supposed to be here? We shouldn't have "ubench" files under op kernel directory. I assume this is in error?

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/ubench_receiver.cpp:
Is this supposed to be here? We shouldn't have "ubench" files under op kernel directory. I assume this is in error?

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/ubench_coordination.cpp:
Is this supposed to be here? We shouldn't have "ubench" files under op kernel directory. I assume this is in error?

SeanNijjar reviewed (changes requested):
Requesting changes here. 

The biggest issue here is the calls into eth dataflow api without idle timers. In general, we want to loop `while (eth_txq_is_busy())` for some amount of time before context switching, otherwise we will lose lots of cycles needlessly. 

IMO the cleanest way to deal with and implement this would be to extend the eth_dataflow_api functions like I did in `ttnn/cpp/ttnn/operations/ccl/kernels/edm/edm_handshake.hpp` for functions like `eth_wait_for_bytes` to accept a time_before_context_switch arg (defaults to 0) so we can just pass in an arg or call an `_busy_wait` version of these functions (rather than polluting the caller code with while loops everywhere).

SeanNijjar commented:
Forgot to add. Please also run a t3000 frequent pipeline as well since that pipeline (and tests inside it) have been modified. You should ensure your tests are also picked up there.

jvegaTT committed:
#0: Making buffer region a multiple of 32

jvegaTT committed:
#0: Merge branch 'jvega/barrier' of github.com:tenstorrent/tt-metal into jvega/barrier

jvegaTT committed:
#12516: Modifying dataflow_api.h for performance

Creating the send_eth_receiver_channnel_done for when the eth_channel_sync_t object is already known
Will spin loop a bit before calling run_routing() for semaphore waits to prevent perf loss from over-calling run_routing()
Formatting changes for comments

jvegaTT committed:
#12516: Applying PR recommendations

jvegaTT committed:
#12516: Removing now depricated test_barrier.py

jvegaTT committed:
#12516: Adding test validation for all gather in sharded test

jvegaTT committed:
#12516: Moving from launch_op to invoke

jvegaTT committed:
#12516: Other suggested changes

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/barrier_op.cpp:
```diff
@@ -0,0 +1,116 @@
+} //namespace ccl end
+
+} //namespace operations end
+} // namespace ttnn end
```
Hey @TT-BrianLiu - can you clarify this please because I still see "register_operation" and a launch op call specified there (it's just nested now in a function that's in another file).

jvegaTT replied:
We need to manually call launch_op as there are slight differences to the launch structure for each of the devices. As much as possible was moved back to the invoke call but there is still a need to modify the default launch_op call. This method was also used in matmul_op.cpp as it also had per-device changes.

jvegaTT committed:
#12516: Reusing pre-made handshake API

jvegaTT commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/host/barrier_full_worker_grid.cpp:
```diff
@@ -0,0 +1,189 @@
+    const uint32_t sample_page_size=16;
+    //Our intro into the kernel code
+    //Turn our tensors into a vector of tensors with 1 entry each
+    std::vector<Tensor> input_tensors = {input_tensor};
```
ttnn::ccl::CCLOpConfig does not give const guarantees on input tensors

SeanNijjar replied:
Thanks. Double checked why this is. You can disregard my comment. Looks like we should update `CCLOpConfig` to not prevent this.

jvegaTT committed:
#12516: Moving as much as possible pre-processing from launch_op to invoke

jvegaTT committed:
#12516: Removed redundant send_eth_receiver_channel_done which is now part of the api

jvegaTT committed:
#12516: Added pyskip for n300

jvegaTT committed:
Merge branch 'main' into jvega/barrier

jvegaTT committed:
#0: Removing blank space in C make lists

jvegaTT committed:
#0: Merge branch 'jvega/barrier' of github.com:tenstorrent/tt-metal into jvega/barrier

SeanNijjar commented on tt_metal/hw/inc/ethernet/dataflow_api.h:
```diff
@@ -439,18 +461,24 @@ void eth_receiver_done() {
         0,
-        ((uint32_t)(&(erisc_info->channels[channel].bytes_sent))) >> 4,
-        ((uint32_t)(&(erisc_info->channels[channel].bytes_sent))) >> 4,
+        ((uint32_t)(channel_sync)) >> 4,
```
Although bytes_sent is at the start of the  struct, you should still take the address of the field so that the code will work with any struct changes.

jvegaTT replied:
Was just trying to avoid that extra de-reference for perf but I guess a good optimizer will realize that it can be optimized away

SeanNijjar replied:
There is no derefence required since we're simply taking the address. (See here as an example: https://godbolt.org/z/91Y8E7n18). Alternatively, you could do `static_assert(offsetof(eth_channel_sync_t, bytes_sent) == 0), "assumption failed yada yada yada");`

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/barrier_op.cpp:
```diff
@@ -0,0 +1,98 @@
+    std::optional<chip_id_t> receiver_device_id = std::nullopt;
+    std::optional<chip_id_t> sender_device_id = std::nullopt;
+    for (uint32_t i = 0; i < num_devices; ++i) {
+        if (devices.at(i) == input_tensor.device()) {
```
Note for future: we'll need to update this to use the factored out function that @Aswinmcw is pulling out so this device selection logic is using the commonized implementation.

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/barrier_op.cpp:
```diff
@@ -0,0 +1,98 @@
+
+void Barrier::validate(const std::vector<Tensor>& input_tensors) const {
+    //Validate the input tensor
+    TT_FATAL(this->topology == ccl::Topology::Ring, "We currently only support Ring topologies on this OP");
```
" ...barrier op"

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/barrier_op.cpp:
```diff
@@ -0,0 +1,98 @@
+            {
+                const Tensor& input_tensor = input_tensors.at(0);
+                // need to copy and update barrier struct for this particular tensor
+                ttnn::Barrier new_barrier_struct = barrier_struct;
```
Can you please explain this sequence? I'm a bit confused by it. 

It looks like we call `barrier(inpout_tensor, Barrier(..., input_tensor.get_workers())` then we have to manually update the barrier struct. Two things:

1) `barrier_struct` is captured by value/copy here so we shouldn't need this explicit copy
2) Why not just initialize the structure like this `update_structure` call but just do it in the constructor?

jvegaTT replied:
I think my comment may be misleading. If I bring into the invoke function just a variable rather than a constant, I will see corruption between the different devices (the values seem to get overwritten for each subsequent device so in runtime I have the wrong config being deployed). I suspect that internally there must be some form of shallow copying and every time this function is called for each of the devices, we may be overwritting the same memory structure which can lead to a race condition if the device has not loaded in that data when the compiler runs launch_op for a different device. Since it is a constant, I need to be internally creating a copy of the structure that isn't constant so I can update it with the device specific stuff.

jvegaTT replied:
Now I guess I could improve compilation time by a few nanoseconds by instead creating a different barrier_parameters structure that only contains the non device-specific parameters, passing that in, and then combining update_structure with barrier structure initialization which would take in this barrier_parameters structure as an input. However I personally think the extra copying time of doing it this way rather than copying in the leaner barrier_parameters does not outweigh the cost of making this flow harder to follow. And since it is on the compile and not runtime side I didn't think a few nanoseconds were significant, but if you disagree I can do that, it would only take a couple minutes.

SeanNijjar replied:
Thanks I see that the passed arg is const ref. okay sounds good. Don't bother splitting - I was just asking for clarification here.

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/host/barrier_full_worker_grid.cpp:
```diff
@@ -0,0 +1,162 @@
+    //Configure operational parameters
+    auto const& op_config =ttnn::ccl::CCLOpConfig(input_tensors, output_tensors, topology);
+    //Get the configuration file, works for both sharded or unsharded
+    const std::unique_ptr<ttnn::ccl::CclOpTensorConfig> input_tensor_config =
```
dead code?

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/host/barrier_full_worker_grid.cpp:
```diff
@@ -0,0 +1,162 @@
+    //Get the configuration file, works for both sharded or unsharded
+    const std::unique_ptr<ttnn::ccl::CclOpTensorConfig> input_tensor_config =
+        ttnn::ccl::CclOpTensorConfig::build_all_gather_tensor_config(input_tensor);
+    const std::unique_ptr<ttnn::ccl::CclOpTensorConfig> output_tensor_config =
```
dead code?

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/barrier_receiver.cpp:
```diff
@@ -0,0 +1,121 @@
+    //In the Ring Start case, we first send the signal to the sender and then we wait on the ethernet
+    noc_semaphore_inc(channel_sem_addr, 1);
+    uint32_t count = 0;
+    while(eth_channel_syncs->bytes_sent == 0) {
```
This can be replaced with the `eth/dataflow_api` call `eth_wait_for_bytes_on_channel` (you'll just need to pull out the `wait_min` arg and pass in.

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/barrier_receiver.cpp:
```diff
@@ -0,0 +1,121 @@
+    )
+{
+    uint32_t count = 0;
+    while(eth_channel_syncs->bytes_sent == 0) {
```
same comment

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/barrier_receiver.cpp:
```diff
@@ -0,0 +1,121 @@
+        }
+    }
+    eth_channel_syncs->bytes_sent = 0;
+    while (eth_txq_is_busy());
```
This will need a `run_routing`escape path. I think it makes sense to add a `wait_for_eth_txq_cmd_space(wait_min)`

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/kernels/barrier_receiver.cpp:
```diff
@@ -0,0 +1,121 @@
+    eth_channel_syncs->bytes_sent = 0;
+    noc_semaphore_inc(channel_sem_addr, 1);
+    count = 0;
+    while (eth_txq_is_busy())
```
same `wait_for_eth_txq_cmd_space` comment

SeanNijjar reviewed (changes requested):
Some minor comments. I think one of the kernels should be cleaned up further.

TT-BrianLiu commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/barrier_op.cpp:
```diff
@@ -0,0 +1,98 @@
+    return std::vector<SimpleShape>(input_tensors.size(), shape);
+}
+
+std::vector<Tensor> Barrier::create_output_tensors(const std::vector<Tensor>& input_tensors) const {
+    //Tensor is unmodified, return what was passed in
+    return input_tensors;
+}
```
Just an FYI, I think we want to move `create_output_tensors` to be fully manual mode for cases like this. This means you shouldn't have to even implement `compute_output_shapes`.

TT-BrianLiu reviewed (approved):
Looks good

jvegaTT committed:
#12516: Improvements to barrier op and dataflow_api.h

Created an eth_wait_for_bytes_on_channel_sync_addr function for when the channel buffer is not one of the defaults in erisc_info->channels
Added wait_for_eth_txq_cmd_space to properly wait for the txq function to finish rather than using an empty while loop
Moved the barrier frequent test to the ccl folder
Improvements on the barrier op using new aforementioned functions

jvegaTT committed:
#12516: Updated run_t3000_frequent_tests.sh with new location of ccl tests

jvegaTT committed:
Merge branch 'main' into jvega/barrier

SeanNijjar reviewed (approved):
Thanks @jvegaTT !

jvegaTT committed:
Merge branch 'main' into jvega/barrier

jvegaTT committed:
Merge branch 'main' into jvega/barrier

jvegaTT committed:
Merge branch 'main' into jvega/barrier

jvegaTT committed:
Merge branch 'main' into jvega/barrier

dmakoviichuk-tt commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/host/barrier_full_worker_grid.cpp:
```diff
@@ -0,0 +1,157 @@
+    return {receiver_kernel_id, sender_kernel_id,sem_init_id};
+}
+
+static std::tuple<std::vector<uint32_t>,std::vector<uint32_t>,std::vector<uint32_t>> get_rt_args(
```
+ @sminakov-tt 
@jvegaTT is it possible for you to use small_vector or std::array?

jvegaTT replied:
I was worried that tt::tt_metal::SetRuntimeArgs when I read it expected std::vector but I just noticed that a variant of that function exists using stl::Span which should work with std::array. I am running tests just in case but this change should be ok.

jvegaTT replied:
It passed the tests

jvegaTT committed:
#12516: Move rt_args from std::vector to std::array

jvegaTT committed:
#0: Merge branch 'jvega/barrier' of github.com:tenstorrent/tt-metal into jvega/barrier

cfjchu commented on ttnn/cpp/ttnn/operations/ccl/barrier/barrier_pybind.cpp:
```diff
@@ -0,0 +1,76 @@
+        Example:
+            >>> full_tensor = torch.randn([1, 1, 256, 256], dtype=torch.bfloat16)
+            >>> input_tensors = torch.chunk(full_tensor, num_devices, dim)
+            >>> physical_device_ids = ttnn.get_t3k_physical_device_ids_ring()
```
The guidance is to do this instead of calling `ttnn.get_t3k_physical_device_ids_ring()`
```
ttnn.open_mesh_device(ttnn.MeshShape(1, 8), mesh_type=ttnn.MeshType.Ring)
```

jvegaTT replied:
I agree, have applied this change

cfjchu commented on ttnn/cpp/ttnn/operations/ccl/barrier/barrier_pybind.cpp:
```diff
@@ -0,0 +1,76 @@
+            >>> physical_device_ids = ttnn.get_t3k_physical_device_ids_ring()
+            >>> mesh_device = ttnn.open_mesh_device(ttnn.MeshShape(1, 8), physical_device_ids=physical_device_ids[:8])
+            >>> tt_input_tensors = []
+            >>> for i, t in enumerate(input_tensors):
```
Can we work directly with `mesh_tensor`? Why are we disaggregating and aggregating?

jvegaTT replied:
This tensor really doesn't do any data manipulation but we want it to be deployed on all devices in the ring. In fact it will not work if you do not do so (you no longer have a full ring). This bit of code divides a dummy tensor between all 8 devices so that they are all scheduled for execution. In a real use case this will likely be coupled with all-gather reduce-scatter or some other OP dividing a tensor among all 8 devices and we can use the same input mesh for this OP

cfjchu replied:
Looking closer at this doc-string looks like a copy-paste for reduce-scatter? where are we actually showing example usage of ttnn.barrier?

jvegaTT replied:
That is because it is currently basically used in the same way. A good future improvement would be to not need an input tensor at all, but that requires a significant rework I think as to my knowledge there is no current flow for an OP without a data tensor that still has the multi-device view. I honestly just need the kernel code to be deployed on all devices in the ring and to be able to use the APIs that tell me which device each ethernet link is connected to, so I need the devices object with no tensor data and I need for the runtime and compilation to understand it needs to be compiled and launched on all devices. This will likely break a lot of things to get working. On a perf side though the user is likely already building this tensor for another OP in the flow, barrier is unlikely to be used on its own, and can just feed in their input tensor to the barrier one which means I don't see it as a big issue.

cfjchu replied:
I'm making a observation that your pybind documentation for `ttnn.barrier` operation does not actually demonstrate usage of the operation itself.

jvegaTT replied:
Oh ... lol thanks for catching that

cfjchu commented on tech_reports/Programming Mesh of Devices/Programming Mesh of Devices with TT-NN.md:
```diff
@@ -139,7 +140,7 @@ torch_tensor[..., 32:64] = 2.0
 mesh_tensor: ttnn.Tensor = ttnn.from_torch(
     torch_tensor,
-    mesh_mapper=ttnn.ShardTensorToMesh(device_mesh, dim=3),
+    mesh_mapper=ttnn.ShardTensorToMesh(mesh_device, dim=3),
```
thank you!

cfjchu commented on tests/ttnn/unit_tests/operations/ccl/test_barrier_t3000_frequent.py:
```diff
@@ -0,0 +1,362 @@
+    # Use Async mode based on test input config
+    device.enable_async(enable_async)
+    for i, t in enumerate(input_tensors):
+        tt_input_tensors.append(ttnn.Tensor(t, input_dtype).to(layout).to(device.get_devices()[i], mem_config))
```
see my other comment

jvegaTT replied:
Applied suggestion to all instances of the above

cfjchu commented:
We have a `ttnn.synchronize_devices(mesh_device)` API.. Will this be a cheaper alternative then? Or is the usage not entirely overlapping?

jvegaTT committed:
#12516: changing get_t3k_physical_devices_ids_ring to open_mesh_device as recommended

jvegaTT commented:
> We have a `ttnn.synchronize_devices(mesh_device)` API.. Will this be a cheaper alternative then? Or is the usage not entirely overlapping?
synchronize_devices is a host operation which has significant latency in going through the host stack. We then also need to call it for each of the devices one by one before re-issuing new OPs for execution. Barrier on the other hand is a synchronization stage among the devices with no host involvement and it can be chain executed with other OPs (e.g. say a user want to do a bunch of matmuls in parallel to a memory buffer, then have a barrier to ensure they are all done, and then do a reduce scatter on the results and say that needs to be repeated, the whole flow including the synchronization can be compiled together and quickly executed in contrast with needing to reach this host synchronization point to restart deployment after each iteration and have all of that overhead).

jvegaTT committed:
#12516: Applying api recommendations

Using ttnn.synchronize_devices instead of synchronize_device

Using from_torch to create the tensor mesh in one step

jvegaTT commented on tests/ttnn/unit_tests/operations/ccl/test_barrier_t3000_frequent.py:
```diff
@@ -0,0 +1,362 @@
+    # Use Async mode based on test input config
+    device.enable_async(enable_async)
+    for i, t in enumerate(input_tensors):
+        tt_input_tensors.append(ttnn.Tensor(t, input_dtype).to(layout).to(device.get_devices()[i], mem_config))
```
Applied suggestion to all instances of the above

jvegaTT commented on tests/ttnn/unit_tests/operations/ccl/test_barrier_t3000_frequent.py:
```diff
@@ -0,0 +1,362 @@
+        topology=all_gather_topology,
+    )
+    for d in device.get_devices():
+        ttnn.synchronize_device(d)
```
Applied suggestion to all instances of the above

jvegaTT committed:
#12516: Minor documentation error on barrier OP

jvegaTT committed:
Merge branch 'main' into jvega/barrier

cfjchu reviewed (approved):
Looks great!

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/ccl/barrier/barrier.hpp:
```diff
@@ -0,0 +1,26 @@
+
+#include "ttnn/decorators.hpp"
+
+#include "ttnn/operations/reduction/generic/generic_reductions.hpp"
+
+#include "ttnn/cpp/ttnn/operations/ccl/ccl_host_types.hpp"
+
+#include "ttnn/cpp/ttnn/operations/ccl/barrier/device/barrier_op.hpp"
```
can we please remove unnecessary includes? 
I guess you only need one for `ttnn::ccl::Topology`

jvegaTT replied:
Agreed, thanks!

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/ccl/barrier/barrier.cpp:
```diff
@@ -0,0 +1,27 @@
+    ttnn::ccl::Topology topology)
+{
+    MemoryConfig out_memory_config = memory_config.value_or(input_tensor.memory_config());
+    return barrier
```
I guess you call a function from barrier_op.hpp here, but it rather looks like a call to `constexpr auto barrier = ttnn...` this thing.

jvegaTT replied:
I agree that the function would likely be better put in barrier.hpp but as I said in another comment, I really want to maintain consistency in the folder structure among ccl ops just to make navigating and debugging them easier in the future, and this is the way it is done in the others.

ayerofieiev-tt replied:
I am saying there is a clash. You registered your op as `barrier` and you have a function called `barrier`, same namespace. This is not good.

jvegaTT replied:
Oh I see what you mean, yeah will rename the namespace

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/host/barrier_full_worker_grid.cpp:
Why is this file inside "host" folder?
It is not applicable.
This is looks like a program factory for a TT device, not cpu.

jvegaTT replied:
This is still host code which orchestrates the compilation and launching of the OP, but it is running inside of a CPU still. Only the code in the kernels directory is running on a TT device. In terms of folder structure it matches what is done in other CCL OPs

SeanNijjar replied:
I think this is carried over from reduce-scatter. All-gather doesn't use this folder but I wanted to originally signify that the code there was actually intended to run on host (as opposed to device only or both). I think as a future item we should rename the directory to "program_factory" or something along those lines.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/ccl/barrier/device/barrier_op.hpp:
```diff
@@ -0,0 +1,53 @@
+        const std::vector<Tensor> &input_tensors, std::vector<Tensor> &output_tensors) const;
+};
+
+namespace ccl::barrier::detail {
+//Template for the barrier_with_workers function
+//Found in device/host/barrier_full_worker_grid.cpp
+operation::ProgramWithCallbacks barrier_with_workers(
+    const Tensor& input_tensors,
+    const Tensor& output_tensors,
+    const bool is_starting_core,
+    const uint32_t ring_size,
+    const uint32_t ring_index,
+    const std::optional<chip_id_t> receiver_device_id,
+    const std::optional<chip_id_t> sender_device_id,
+    ttnn::ccl::Topology topology);
+}; // namespace ccl::barrier::detail
+
+namespace operations::ccl{
+    //Template for the barrier tensor found in device/barrier_op.cpp
+Tensor barrier(
+    const Tensor& input_tensor,
+    const ttnn::Barrier& barrier_struct);
+} // namespace operations::ccl
```
should these rather be moved to barrier.cpp?

jvegaTT replied:
I could move them. It has no perf impact either way but this matches how it is done in other CCL OPs and I would rather maintain consistency across them.

jvegaTT committed:
#12516: removing unused libraries in barrier.hpp

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/ccl/barrier/barrier.hpp:
```diff
@@ -0,0 +1,22 @@
+
+#include "ttnn/cpp/ttnn/operations/ccl/ccl_host_types.hpp"
+
+#include "ttnn/cpp/ttnn/operations/ccl/barrier/device/barrier_op.hpp"
```
This include must be moved to cpp. It really does not belong in this header. 
Presence of such includes makes our build times very long.

jvegaTT replied:
Agreed, have replaced it with just the decorators library that was needed

jvegaTT committed:
#12516: Removing barrier_op.hpp inclusion in barrier.hpp

jvegaTT committed:
#12516: Changing function name from barrier to barrier_function

jvegaTT committed:
Merge branch 'main' into jvega/barrier

jvegaTT committed:
#12516: Correcting clang-diagnostics-error

jvegaTT committed:
#0: Merge branch 'jvega/barrier' of github.com:tenstorrent/tt-metal into jvega/barrier

jvegaTT committed:
#12516: Other clang errors fix

jvegaTT committed:
Merge branch 'main' into jvega/barrier
